freshmaker/handlers/koji/rebuild_flatpak_application_on_module_ready.py

```diff
@@ -167,3 +179,152 @@ def _filter_images_with_higher_rpm_nvr(self, event, rebuild_images_list):
         )
         images = lb.get_images_by_nvrs(rebuild_images_list, rpm_nvrs=errata_rpm_nvrs)
         return images
+
+    def _get_module_name_stream_set(self, event):
+        """
+        Get set of modules with format NAME:STREAM in an advisory.
+
+        :param event FlatpakModuleAdvisoryReadyEvent: The event this handler
+            is currently handling.
+        :return: a set with item NAME:STREAM
+        :rtype: set
+        """
+        errata = Errata()
+        module_nvrs = errata.get_cve_affected_build_nvrs(event.advisory.errata_id, True)
+
+        module_name_stream_set = set()
+        with koji_service(
+            conf.koji_profile, log, login=False, dry_run=self.dry_run
+        ) as session:
+            for module_nvr in module_nvrs:
+                mmd = session.get_modulemd(module_nvr)
+                name = mmd.get_module_name()
+                stream = mmd.get_stream_name()
+                module_name_stream_set.add(name + ":" + stream)
+
+        return module_name_stream_set
+
+    def _prepare_data_for_compose(
+        self, original_odcs_compose_ids, module_name_stream_set
+    ):
+        """
+        Prepare data before recording compose data in build or
+        creating a new compose and linking it to a build.
+
+        :param original_odcs_compose_ids list: A list of compose ids
+            in a image.
+        :param module_name_stream_set set: A set of module's metadata
+            in format NAME:STREAM of an advisory.
+        :return: a tuple of compose data. The first item is a set of
+            compose ids which will be renewed in build, the second item
+            is a set of required data for creating new composes
+        :rtype: tuple
+        """
+        renewed_odcs_compose_set = set()
+        new_odcs_compose_set = set()
+        for compose_id in original_odcs_compose_ids:
+            compose = self.odcs.get_compose(compose_id)
+            source_type = compose.get("source_type")
+            # source_type of module composes is 2
+            if source_type != 2:
+                renewed_odcs_compose_set.add(compose_id)
+                continue
+
+            name_stream_set = set()
+            source_value = compose.get("source", "")
+            for source in source_value.split():
+                group = source.strip().split(":")[:2]
+                name_stream_set.add(group[0] + ":" + group[1])
+
+            # If all the sources are not in advisory, renew them
+            if len(name_stream_set - module_name_stream_set) == len(name_stream_set):
+                renewed_odcs_compose_set.add(compose_id)
+            else:
+                # If some source in advisory, create composes for all separately
+                new_odcs_compose_set.update(name_stream_set)
```

Comment:
I believe odcs module compose source should be module nsvc instead of just name and stream, can you check this?

Reply:
Yes. When I use method GET, I can see the source format is nvsc, but I checked the [doc](https://pagure.io/odcs/#:~:text=Requesting%20new%20compose), when we create a new compose, for module source type, the source can be "White-space separated NAME:STREAM or NAME:STREAM:VERSION of modules to include in compose." I am wondering if the VERSION should be created automatically.

Reply:
If you request compose with only module `name` + `stream`, I guess ODCS should get the latest modules from that `name` and `stream`, which can be different from the modules in the trigger advisory, so I believe at least `version` is also required, but I'm not sure about `context`, I'd suggest checking with ODCS team if that is not documented.

Reply:
Ok. Checked with ODCS people and as you mentioned name + stream get the latest modules. Adding version as well